core/services/gateway/handlers/functions/orm.go

```diff
@@ -0,0 +1,136 @@
+type cachedSubscriptionRow struct {
+	SubscriptionID        uint64
+	Balance               string
+	Owner                 common.Address
+	BlockedBalance        string
+	ProposedOwner         common.Address
+	Consumers             pq.ByteaArray
+	Flags                 []uint8
+	RouterContractAddress common.Address
+}
+
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig, routerContractAddress common.Address) (ORM, error) {
+	if db == nil || cfg == nil || lggr == nil || len(routerContractAddress) == 0 {
+		return nil, ErrInvalidParameters
+	}
+
+	return &orm{
+		q:                     pg.NewQ(db, lggr, cfg),
+		routerContractAddress: routerContractAddress,
+	}, nil
+}
+
+func (o *orm) GetSubscriptions(offset, limit uint, qopts ...pg.QOpt) ([]CachedSubscription, error) {
+	var cacheSubscriptions []CachedSubscription
+	var cacheSubscriptionRows []cachedSubscriptionRow
+	stmt := fmt.Sprintf(`
+		SELECT subscription_id, owner, balance, blocked_balance, proposed_owner, consumers, flags, router_contract_address
+		FROM %s
+		WHERE router_contract_address = $1
+		ORDER BY subscription_id DESC
+		OFFSET $2
+		LIMIT $3;
+	`, tableName)
+	err := o.q.WithOpts(qopts...).Select(&cacheSubscriptionRows, stmt, o.routerContractAddress, offset, limit)
+	if err != nil {
+		return cacheSubscriptions, err
+	}
+
+	for _, cs := range cacheSubscriptionRows {
+		consumers := make([]common.Address, 0)
+		for _, csc := range cs.Consumers {
+			consumers = append(consumers, common.BytesToAddress(csc))
+		}
+
+		balance, ok := big.NewInt(0).SetString(cs.Balance, 10)
+		if !ok {
+			return nil, fmt.Errorf("error parsing balance %s", cs.Balance)
+		}
+
+		blockedBalance, ok := big.NewInt(0).SetString(cs.BlockedBalance, 10)
+		if !ok {
+			return nil, fmt.Errorf("error parsing blockedBalance %s", cs.BlockedBalance)
+		}
+
+		cacheSubscriptions = append(cacheSubscriptions, CachedSubscription{
+			SubscriptionID: cs.SubscriptionID,
+			IFunctionsSubscriptionsSubscription: functions_router.IFunctionsSubscriptionsSubscription{
+				Balance:        balance,
+				Owner:          cs.Owner,
+				BlockedBalance: blockedBalance,
+				ProposedOwner:  cs.ProposedOwner,
+				Consumers:      consumers,
+				Flags:          [32]byte(cs.Flags),
+			},
+		})
+	}
+
+	return cacheSubscriptions, nil
+}
+
+func (o *orm) UpsertSubscription(subscription CachedSubscription, qopts ...pg.QOpt) error {
+	stmt := fmt.Sprintf(`
+		INSERT INTO %s (subscription_id, owner, balance, blocked_balance, proposed_owner, consumers, flags, router_contract_address)
+		VALUES ($1,$2,$3,$4,$5,$6,$7,$8) ON CONFLICT (subscription_id, router_contract_address) DO UPDATE
+		SET owner=$2, balance=$3, blocked_balance=$4, proposed_owner=$5, consumers=$6, flags=$7, router_contract_address=$8;`, tableName)
+
+	if subscription.Balance == nil {
+		subscription.Balance = big.NewInt(0)
+	}
+
+	if subscription.BlockedBalance == nil {
+		subscription.BlockedBalance = big.NewInt(0)
+	}
+
+	_, err := o.q.WithOpts(qopts...).Exec(
```

Comment:
Does this work correctly when a sub is cancelled? Can we add a test case please?

Reply:
added